util/types/datum.go

```diff
@@ -1327,6 +1327,8 @@ func (d *Datum) toSignedInteger(sc *variable.StatementContext, tp byte) (int64,
 	case KindString, KindBytes:
 		iVal, err := StrToInt(sc, d.GetString())
 		if err != nil {
+			// adjust the overflow value
+			iVal, _ = ConvertIntToInt(iVal, lowerBound, upperBound, tp)
```

Comment:
I'm sorry,  I forgot `StrToInt` implementation details.
You can consider it like the [logical](https://github.com/pingcap/tidb/pull/3508/files#diff-b67a235702396fe5cc8ec3d67962697eR1366).@louishust